### PR TITLE
linear_congruential_engine: add using more precision to prevent overflow

### PR DESCRIPTION
**Pull request by LRFLEW**
This PR is a followup to #81080, ~~and as such includes the commit from that PR. This should either be merged after that one is merged (and this one is rebased), or this one can take the place of that PR. Since the other PR has gotten pretty far in terms of reviews, and this PR is more involved, the other PR should probably be merged before this one.~~

This PR makes two major changes to how the LCG operation is computed:

The first is that I added an additional case where `ax + c` might overflow the intermediate variable, but `ax` by itself won't. In this case, it's much better to use `(ax mod m) + c mod m` than the previous behavior of falling back to Schrage's algorithm. The addition modulo is done in the same way as when using Schrage's algorithm (i.e. `x += c - (x >= m - c)*m`), but the multiplication modulo is calculated directly, which is faster.

The second is that I added handling for the case where the `ax` intermediate might overflow, but Schrage's algorithm doesn't apply (i.e. r > q). In this case, the only real option is to increase the precision of the intermediate values. The good news is that - for `x`, `a`, and `c` being n-bit values - `ax + c` will never overflow a 2n-bit intermediary, meaning this promotion can only happen once, and will always be able to use the simplest implementation. This is already the case for 16-bit LCGs, as libcxx chooses to compute them with 32-bit intermediate values. For 32-bit LCGs, I simply added code similar to the 16-bit case to use the existing 64-bit implementations. Lastly, for 64-bit LCGs, I wrote a case that calculates it using `unsigned __int128` if it is available to use.

While this implementation covers a *lot* of the missing cases from #81080, this still won't compile **every** possible `linear_congruential_engine`. Specifically, if `a`, `c`, and `m` are chosen such that it needs 128-bit integers, but the platform doesn't support `__int128` (eg. 32-bit x86), then it will fail to compile. However, this is a fairly rare case to see actually used, and libcxx would be in good company with this, as [libstdc++ also fails to compile under these circumstances](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=87744). Fixing **this** gap would require even **more** work of further complexity, so that would probably be best handled by a different PR (I'll put more details on what that PR would entail in a comment).

### CODE REVIEW

**Comment by LRFLEW** [comment]
Ok, I'm happy with this PR now, so I'm gonna remove the Draft status. The CI is giving me issues, but it seems like it's passing all the cases at this point. Pretty much the only thing I'm not 100% happy with is the naming of the new functions in the test code, but it should be fine. Let me know if you have any feedback or suggestions on this.

File: libcxx/test/std/numerics/rand/rand.eng/rand.eng.lcong/assign.pass.cpp [hunk 1]
```diff
@@ -61,24 +61,33 @@ test()
   test1<T, A, 0, M>();
   test1<T, A, M - 2, M>();
   test1<T, A, M - 1, M>();
+}
+
+template <class T>
+void test_ext() {
+  const T M(static_cast<T>(-1));
 
-  /*
```
Comment (mordante) [inline libcxx/test/std/numerics/rand/rand.eng/rand.eng.lcong/assign.pass.cpp:65]: I assume the next line's " (not implemented)" is no longer true too. (The same for similar enabled tests).

File: libcxx/test/std/numerics/rand/rand.eng/rand.eng.lcong/alg.pass.cpp [hunk 2]
```diff
@@ -38,12 +38,12 @@ int main(int, char**)
 
     // m might overflow. The overflow is not OK and result will be in bounds
     // so we should use Schrage's algorithm
-    typedef std::linear_congruential_engine<T, (1ull << 32), 0, (1ull << 63) + 1> E2;
+    typedef std::linear_congruential_engine<T, 0x100000000ull, 0, (1ull << 63) + 1ull> E2;
```
Comment (mordante) [inline libcxx/test/std/numerics/rand/rand.eng/rand.eng.lcong/alg.pass.cpp:41]: Why did you change `(1ull << 32)` to `0x100000000ull`?

**Review by mordante — COMMENTED** [review]
Thanks for working on this! I mainly took a quick look at the patch. I hope to have time to do a better review soon. (I really need to look at how the algorithm should work in the first place before reviewing.)

File: libcxx/test/std/numerics/rand/rand.eng/rand.eng.lcong/alg.pass.cpp (hunk 2 above)
Comment (LRFLEW) [reply libcxx/test/std/numerics/rand/rand.eng/rand.eng.lcong/alg.pass.cpp:41]: No *specific* reason. Mostly, I was trying to make this code a little cleaner and consistent. I figured that the `(1ull << 63) - 0x1ffffffffull` I added a few lines down would be longer and less clear if written as `(1ull << 63) - (1ull << 33) + 1ull`, and if I'm writing the 31-bit hex literal there, then I might as well do it here as well. I don't mind changing this back if that's what's preferred, though.

Comment (mordante) [reply libcxx/test/std/numerics/rand/rand.eng/rand.eng.lcong/alg.pass.cpp:41]: To me `1ull << 32` is easier to parse. With `0x100000000ull` I need to count the zeros. (Since the code is used in C++11 we can't write `0x1'0000'0000ull`.) So personally I really like the shifted version better.

**Comment by mordante** [comment]
FYI I've not had time to look at this yet, I still hope to find time for it soon.

**Review by mordante — COMMENTED** [review]
Thanks for the patch! I finally found time to have a good look at it. I think it's mostly good. Can you update the commit message; it mentions the patch is WIP.

I'm happy to accept the patch and have a compile-time error when 128-bit integrals are missing and it's needed. We can add that in a later patch.

File: libcxx/include/__random/linear_congruential_engine.h [hunk 3]
```diff
@@ -80,21 +97,40 @@ struct __lce_ta<__a, 0, __m, (unsigned long long)(~0), true> {
...
   _LIBCPP_HIDE_FROM_ABI static result_type next(result_type __x) { return __a * __x + __c; }
 };
 
 // 32
 
+template <unsigned long long __a, unsigned long long __c, unsigned long long __m>
+struct __lce_ta<__a, __c, __m, unsigned(~0), 3> {
+  typedef unsigned result_type;
+  _LIBCPP_HIDE_FROM_ABI static result_type next(result_type __x) {
+    return static_cast<result_type>(__lce_ta<__a, __c, __m, (unsigned long long)(~0)>::next(__x));
```
[the remark this answers is not in the document]
Comment (LRFLEW) [reply libcxx/include/__random/linear_congruential_engine.h:128]: I was thinking this looked weird, but this is what was there before. I agree it looks wrong (even if it technically isn't), though. My issue with this change is that it can't *quite* be applied universally. Even if I changed this in all of the specializations of `__lce_ta`, the definition of `linear_congruential_engine::_Mp` would still have that issue, as it's currently defined as `result_type(~0)`.

Personally, I'd rather change this line to `~(unsigned long long)(0)`, as I think it's clearer what the intent is, and is less prone to mistakes from using the wrong suffix.

File: libcxx/include/__random/linear_congruential_engine.h [hunk 4]
```diff
@@ -30,28 +30,45 @@ template <unsigned long long __a,
...
 struct __lce_ta;
 
 // 64
 
+#ifndef _LIBCPP_HAS_NO_INT128
+template <unsigned long long _Ap, unsigned long long _Cp, unsigned long long _Mp>
+struct __lce_ta<_Ap, _Cp, _Mp, (unsigned long long)(~0), 3> {
+  typedef unsigned long long result_type;
+  _LIBCPP_HIDE_FROM_ABI static result_type next(result_type __xp) {
+    __extension__ typedef unsigned __int128 calc_type;
```
[the remark this answers is not in the document]
Comment (LRFLEW) [reply libcxx/include/__random/linear_congruential_engine.h]: I used `typedef` here mostly because this header uses `typedef` in a number of places (eg. all the `result_type` typedefs), most likely due to its origin as a TR1 header. I went ahead and changed this one instance to use `using` because I do think it looks better, but haven't touched any of the other `typedef` statements in this header.

Similarly, I hadn't __uglified the type name due to the un-uglified `result_type` type definitions for the `__lce_ta` specializations. I went ahead and made that change for this type definition, though.

File: libcxx/include/__random/linear_congruential_engine.h [hunk 5]
```diff
@@ -30,28 +30,45 @@ template <unsigned long long __a,
...
           unsigned long long __c,
           unsigned long long __m,
           unsigned long long _Mp,
-          bool _UseSchrage = __lce_alg_picker<__a, __c, __m, _Mp>::__use_schrage>
+          int _Mode = __lce_alg_picker<__a, __c, __m, _Mp>::__mode>
 struct __lce_ta;
 
 // 64
 
+#ifndef _LIBCPP_HAS_NO_INT128
```
[the remark this answers is not in the document]
Comment (LRFLEW) [reply libcxx/include/__random/linear_congruential_engine.h:62]: The static assert in `__lce_alg_picker` is intended to provide the "user understandable" error message for this case, and is based on the static assert was there before. I am trying to get the full output from triggering the static assert, but I don't have multilib setup completely on my system, so it might take me a bit to get that to you.

**Comment by LRFLEW** [comment]
I addressed some of the review comments, and responded to others with followup questions / suggestions.

> Can you update the commit message; it mentions the patch is WIP.

I'm not seeing that. The PR had "[WIP]" at one point, but I don't see that in the commit message itself. Could you clarify what exactly you want me to change?

File: libcxx/include/__random/linear_congruential_engine.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (LRFLEW) [reply libcxx/include/__random/linear_congruential_engine.h:66]: Ok, I got the multilib tests working. I went and commented out the `#ifndef _LIBCPP_HAS_NO_INT128` in assign.pass.cpp, and this is how `llvm_lit` reported the first compile error with clang++:

```
# | In file included from llvm-project/libcxx/test/std/numerics/rand/rand.eng/rand.eng.lcong/assign.pass.cpp:16:
# | In file included from llvm-project/build/include/c++/v1/random:1686:
# | In file included from llvm-project/build/include/c++/v1/__random/default_random_engine.h:13:
# | llvm-project/build/include/c++/v1/__random/linear_congruential_engine.h:52:17: error: static assertion failed due to requirement '18446744073709551615ULL != (unsigned long long)(~0) || false || false || false': The current values for a, c, and m are not currently supported on platforms without __int128
# |    52 |   static_assert(_Mp != (unsigned long long)(~0) || _Full || _Part || _Schrage,
# |       |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# | llvm-project/build/include/c++/v1/__random/linear_congruential_engine.h:61:34: note: in instantiation of template class 'std::__lce_alg_picker<18446744073709551613, 0, 18446744073709551615, 18446744073709551615>' requested here
# |    61 |           __lce_alg_type _Mode = __lce_alg_picker<__a, __c, __m, _Mp>::__mode>
# |       |                                  ^
# | llvm-project/build/include/c++/v1/__random/linear_congruential_engine.h:285:44: note: in instantiation of default argument for '__lce_ta<18446744073709551613ULL, 0ULL, 18446744073709551615ULL, 18446744073709551615ULL>' required here
# |   285 |     return __x_ = static_cast<result_type>(__lce_ta<__a, __c, __m, _Mp>::next(__x_));
# |       |                                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
# | llvm-project/libcxx/test/std/numerics/rand/rand.eng/rand.eng.lcong/assign.pass.cpp:30:11: note: in instantiation of member function 'std::linear_congruential_engine<unsigned long long, 18446744073709551613, 0, 18446744073709551615>::operator()' requested here
# |    30 |     (void)e1();
# |       |           ^
# | llvm-project/libcxx/test/std/numerics/rand/rand.eng/rand.eng.lcong/assign.pass.cpp:71:3: note: in instantiation of function template specialization 'test1<unsigned long long, 18446744073709551613ULL, 0ULL, 18446744073709551615ULL>' requested here
# |    71 |   test1<T, M - 2, 0, M>();
# |       |   ^
# | llvm-project/libcxx/test/std/numerics/rand/rand.eng/rand.eng.lcong/assign.pass.cpp:90:5: note: in instantiation of function template specialization 'test_ext<unsigned long long>' requested here
# |    90 |     test_ext<unsigned long long>();
# |       |     ^
```

File: libcxx/include/__random/linear_congruential_engine.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (mordante) [reply libcxx/include/__random/linear_congruential_engine.h:139]: I'm happy with -1, assuming the compiler will not warn about it. The reason why there where no objections in a later standard (not sure which version) to require two's complement math was that all hardware platforms use two's complement; so I'm happy with the assumption of two's complement.

**Review by mordante — APPROVED** [review]
> I addressed some of the review comments, and responded to others with followup questions / suggestions.
> 
> > Can you update the commit message; it mentions the patch is WIP.
> 
> I'm not seeing that. The PR had "[WIP]" at one point, but I don't see that in the commit message itself. Could you clarify what exactly you want me to change?

At the end of the commit message
"
I currently consider this PR to be a WIP because a) I haven't gotten the test cases written properly to avoid failing when __int128 isn't available, and b) I'm not 100% sure about how I've structured / formatted the changes, and may still want to tweak it before merging. I'm making the PR now so I can start getting feedback if anybody has any.
"

Other than the last comments the patch LGTM! Thanks for your work and patience.

**Comment by LRFLEW** [comment]
Double-checking this, I realized I forgot to respond to the last message. That message you're referring to was not in the commit message, but only in the PR description. I went ahead and removed it anyways, though. I've addressed all the review comments, so unless there's anything else, this should be ready to merge.

**Comment by mordante** [comment]
> Double-checking this, I realized I forgot to respond to the last message. That message you're referring to was not in the commit message, but only in the PR description. I went ahead and removed it anyways, though. I've addressed all the review comments, so unless there's anything else, this should be ready to merge.

FYI When we merge in GitHub, by default, the PR description is the commit message.
